### PR TITLE
Removing typings dependency and adding just @types/mocha.

### PR DESCRIPTION
**Pull request by EdwinGuzman**
Also updating npm `prepublish` to `prepublishOnly`. Using `prepublishOnly` so it won't build the project when running `npm install`.

Removed the `typings` dependency and was going to add `@types` support but it seems it's only needed for mocha. This allows the global reference for mocha that use to be in the unit test to be removed, as it's now in tsconfig.json.

### CODE REVIEW

**Comment by aslagle** [comment]
I'm really confused that this works. Aren't typings still needed for other dependencies like React?

**Comment by EdwinGuzman** [comment]
Yeah... I think there's a dependency which has `@types/react` as its own dependency and it brings it into the project. I should bring them in directly instead of depending on those dependencies.

**Comment by aslagle** [comment]
Tests still pass for me when I do the `React.ABCD` thing on this branch, so I think it's not using those types. I don't know why Typescript doesn't complain.

**Comment by aslagle** [comment]
It seems like it will work to add "@types/react": "^15.0.20", and I think that will let you get rid of all the extra type declarations for state, props, etc. It gives you a bunch of type errors, but all look fixable - mostly changing Element to JSX.Element in a whole bunch of places.

**Comment by EdwinGuzman** [comment]
Errors like the following?
```
__tests__/components/Fieldset-test.tsx(13,7): error TS2605: JSX element type 'Fieldset' is not a constructor function for JSX elements.
__tests__/components/Fieldset-test.tsx(13,17): error TS2322: Type '{ legend: string; elements: Element[]; }' is not assignable to type 'IntrinsicAttributes & IntrinsicClassAttributes<Fieldset> & FieldsetProps'.
...
``` 
I decided to try @types/react 16.4.7 since it seems that corresponds to ts 2.7 according to https://www.npmjs.com/package/@types/react. They seem fixable but I could not find a solution for even the first one...

**Comment by aslagle** [comment]
Yeah. For that one you have to change the return type of `render` to be `JSX.Element` instead of `Element`, and the type of `elements` in the props to also use `JSX.Element` instead of `Element`.

**Comment by EdwinGuzman** [comment]
Cool. I'll fix them later today if I get a chance.

**Comment by EdwinGuzman** [comment]
@aslagle this is ready for another look. Added @types/node and @types/react to remove the props, state, and global declarations. A couple of things: I noticed that it's not common to have `class ... extends React.Component<{}, {}>` instead of `Component<void, void>`. I also had to cast the style object `as React.CSSProperties` but since that's in the `Demo` it doesn't actually affect the source components.

File: src/Demo.tsx [hunk 1]
```diff
@@ -14,9 +14,8 @@ export interface DemoState {
   infoText: string;
 }
 
-export default class Demo extends React.Component<any, DemoState> {
+export default class Demo extends React.Component<{}, DemoState> {
   state: DemoState;
```
Comment (aslagle) [inline src/Demo.tsx]: You should be able to remove this line too.

**Comment by aslagle** [comment]
I'm not sure what you mean by it's not common. `Component<void, void>` is explicitly saying the component has no state and no props. `Component<{}, {}>` is saying the state and props can be anything. If the component does in fact have no state and no props, I think it's better to use `void`.

File: src/Demo.tsx (hunk 1 above)
Comment (EdwinGuzman) [reply src/Demo.tsx]: Missed this. Updated.

**Comment by EdwinGuzman** [comment]
What I mean is that it's prevalent in tutorials and in the typescript react documentation. I did not see any use of `Component<void, void>`. I can check again but it seems it's "ok" to not use `void` after 2.6 (I think).

**Comment by aslagle** [comment]
I'd have to look at it but that sounds like people just don't know how to use the types correctly. That was what we did when we started, because it was really hard to understand the react typings. I could understand if they didn't want beginners to have to deal with them, but it takes away some of the advantage of Typescript - if a component's not supposed to have state, it's good to get a type error when you try to access its state.

**Comment by aslagle** [comment]
Ah, `{}` is an empty object rather than `any`, so that makes sense. It's still saying that you can't have any props or state.

**Pull request merged**